### DISCUSSION TITLE
[Retrospect] NLS to ICU transition on Windows in .NET 5

### DISCUSSION THREAD

**Discussion by am11** [dotnet/runtime#43911] [General] [answered]
There are noticeable number of issues being reported due to the NLS to ICU transition on Windows in .NET 5.

When NCL team introduced managed socket handler, to move away from libcurl dependency, an opt-in environment variable `DOTNET_SYSTEM_NET_HTTP_USESOCKETSHTTPHANDLER` was provided. In dotnet 2.0, managed socket handler was shipped but disabled by default. That gave feature some time to improve/mature. Then in later releases, it was turned on by default.

Were there similar considerations for NLS to ICU transition on Windows, to keep it opt-in for a release or two (instead of opt-out, `DOTNET_SYSTEM_GLOBALIZATION_USENLS`)?

**Comment by eerhardt** [comment]
cc @tarekgh

**Comment by tarekgh** [comment, marked as answer]
@am11 

> Were there similar considerations for NLS to ICU transition on Windows, to keep it opt-in for a release or two (instead of opt-out, DOTNET_SYSTEM_GLOBALIZATION_USENLS)?

We spent some time thinking about that and we decided to have ICU as the default. 
- If we have it opt-in instead of opt-out, that means for sure the majority of apps will not opt-in by default or even cares to opt-in. That means if you want to flip the switch at some point you'll end up with the exact same situation. 
- The .NET 5.0 is one of the major releases and it is the best chance you can introduce such changes. Waiting more releases would complicate the transition than today as we expect more apps will be built and migration from .NET Framework will increase. So, the chance to break more apps in the future would be much bigger than today.
- The globalization data changes happens even with NLS especially Windows is converging to CLDR. Look at Windows blog https://docs.microsoft.com/en-us/archive/blogs/shawnste/locale-data-in-windows-10-cldr which telling the same message. So, such complaints is coming anyway. I agree the volume of the complaints could be less but it will still be there.
- ICU is promoted by Windows too. Many apps already started using ICU on Windows. So migration to use ICU is happening everywhere and not just .NET.

CC @jkotas @danmosemsft @ericstj @safern @stephentoub @GrabYourPitchforks

**Comment by jkotas** [comment]
> In dotnet 2.0, managed socket handler was shipped but disabled by default. That gave feature some time to improve/mature. Then in later releases, it was turned on by default.

ICU based globalization has been pretty mature already. We have been shipping it on Unix for last 5 years. We do not expect that it will get significantly better over time. It is one of the factors that made this case different from the earlier socket handler case.
